### PR TITLE
Test: IntervalIndex.astype("category") doesn't preserve exact interval dtype in categories

### LINKED ISSUE #38316: BUG: IntervalIndex.astype("category") doesn't preserve exact interval dtype in categories
Somewhere in the conversion, before factorizing, we convert the interval array/index to a object-dtype numpy array of Interval objects, and so afterwards *infer* the IntervalDtype again when creating the categories.

Example consequence is that if you have uint64 intervals, they get inferred as int64 afterwards:

```python
In [29]: index = pd.IntervalIndex.from_breaks(np.arange(5, dtype="uint64"))

In [30]: index
Out[30]: 
IntervalIndex([(0, 1], (1, 2], (2, 3], (3, 4]],
              closed='right',
              dtype='interval[uint64]')  # <---- unsigned ints

In [31]: pd.CategoricalIndex(index)
Out[31]: CategoricalIndex([(0, 1], (1, 2], (2, 3], (3, 4]], categories=[(0, 1], (1, 2], (2, 3], (3, 4]], ordered=False, dtype='category')

In [33]: pd.CategoricalIndex(index).dtype.categories
Out[33]: 
IntervalIndex([(0, 1], (1, 2], (2, 3], (3, 4]],
              closed='right',
              dtype='interval[int64]')  # <---- no longer uint64
```

### PR DESCRIPTION
**Pull request by weikhor**
- [ ] closes #38316
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -2283,7 +2284,7 @@ def maybe_reorder(
             result_index = None
             if len(arrays) == 0 and index is None and length == 0:
                 # for backward compat use an object Index instead of RangeIndex
-                result_index = Index([])
+                result_index = RangeIndex(start=0, stop=0, step=1)
```
Comment (phofl) [inline pandas/core/frame.py:2287]: This is not part of this pr, correct?

Comment (weikhor) [reply pandas/core/frame.py]: yes. My mistake. I will revert back.

File: pandas/tests/arrays/categorical/test_dtypes.py [hunk 2]
```diff
@@ -134,3 +136,11 @@ def test_iter_python_types_datetime(self):
         cat = Categorical([Timestamp("2017-01-01"), Timestamp("2017-01-02")])
         assert isinstance(list(cat)[0], Timestamp)
         assert isinstance(cat.tolist()[0], Timestamp)
+
+    def test_interval_index_category(self):
+        # GH 38316
+        index = IntervalIndex.from_breaks(np.arange(5, dtype="uint64"))
+
+        result_subtype = CategoricalIndex(index).dtype.categories.dtype.subtype
+        expected_subtype = "uint64"
```
Comment (mroeschke) [inline pandas/tests/arrays/categorical/test_dtypes.py:145]: Could you instead compare the result of `CategoricalIndex(index).dtype.categories` (the `IntervalIndex`)?

Comment (weikhor) [reply pandas/tests/arrays/categorical/test_dtypes.py:145]: Sure, Will update

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @weikhor
